src/qt3utils/applications/controllers/princeton_spectrometer.py

```diff
@@ -0,0 +1,139 @@
+        self.logger = logging.getLogger(__name__)
+        self.logger.setLevel(logger_level)
+
+        self.spectrometer = princeton.Spectrometer()
+        self.last_config_dict = {}
+
+        self.wave_start = None
+        self.wave_end = None
+        self.last_measured_spectrum = None
+        self.last_wavelength_array = None
+
+    @property
+    def clock_rate(self) -> float:
+        try:
+            _t = self.spectrometer.exposure_time / 1000.0  # Converting from milliseconds to seconds.
+        except Exception as e:
+            self.logger.error(e)
+            _t = 2  # TODO: better default behavior. Should this be -1? 1? or should Spectrometer be changed.
+        return 1.0 / _t
+
+    def start(self) -> None:
+        """
+        Nothing to be done in this method. All acquisition is happening in the "sample_spectrum" method.
+        """
+        self.logger.debug('calling QT3ScanPrincetonSpectrometerController start')
+
+    def stop(self) -> None:
+        """
+        Implementations should do necessary steps to stop acquiring data.
+        """
+        #TODO: Need to implement a feature to pause scan here. If there is a way to interrupt data acquistion, do that here. Otherwise, do nothing
+        self.logger.debug('calling QT3ScanPrincetonSpectrometerController stop')
+
+    def close(self) -> None:
+        self.spectrometer.finalize()
+
+    def sample_spectrum(self) -> Tuple[np.ndarray, np.ndarray]:
+        self.last_measured_spectrum, self.last_wavelength_array = (self.spectrometer.acquire_step_and_glue([self.wave_start, self.wave_end]))
+        self.logger.debug(f'acquired spectrum from {self.last_wavelength_array[0]} to {self.last_wavelength_array[-1]} nm')
+        return self.last_measured_spectrum, self.last_wavelength_array
+
+    def configure(self, config_dict: dict) -> None:
+        """
+        This method is used to configure the spectrometer with the provided settings.
+        """
+        self.logger.debug("Calling configure on the Princeton Spectrometer data controller")
+        self.last_config_dict.update(config_dict)
+
+        self.spectrometer.experiment_name = config_dict.get('experiment_name', self.spectrometer.experiment_name)
+        self.spectrometer.center_wavelength = float(config_dict.get('center_wavelength', self.spectrometer.center_wavelength))
+        self.spectrometer.exposure_time = float(config_dict.get('exposure_time', self.spectrometer.exposure_time))
+        self.spectrometer.temperature_sensor_setpoint = float(config_dict.get('temperature_sensor_setpoint', self.spectrometer.temperature_sensor_setpoint))
+        self.spectrometer.num_frames = float(config_dict.get('num_frames', self.spectrometer.num_frames))
+        self.spectrometer.grating = config_dict.get('grating', self.spectrometer.grating)
+        self.wave_start = float(config_dict.get('wave_start', self.wave_start))
+        self.wave_end = float(config_dict.get('wave_end', self.wave_end))
```

Comment:
I agree that we need to edit to stop casting them to floats. I do not think that all of these can be passed in as `int` though, but I will be able to verify this plus the requirement of whole numbers for milliseconds soon.

Reply:
right -- yeah, some of them could be floats and some might be ints. need to figure out which

Reply:
To answer your question regarding the `exposure_time`, it allows fractions of a millisecond and in the documentation it is of type `double` so I will address that

Reply:
@gadamc Should we not be casting at all? I removed casting for everything but Lightfield doesn't like that. So I type casted them, but with the correct types this time. Most are still floats though since that is what is in the documentation...

Reply:
The documentation is a little confusing to me. It seems that they're all strings!  I think we can just try to test this directly and explicitly in a python shell or Jupyter notebook

Reply:
I think there is some confusion. These values are actually floats, I totally understand your confusion though because the documentation is formatted poorly. For example in the image I have linked below it says `GratingSettingWavelength` as `String`, but I assume that is just talking about what type the name of the property is. The "Remarks" section actually has the data type and it is of type `Double.` I have tested these types and they return `floats` and if you do actually use a `String` you will encounter and error.
![image](https://github.com/qt3uw/qt3-utils/assets/19673888/ce34b4bb-2391-4b6b-bb3d-86ca5e8a66d5)